driver-core/src/main/com/mongodb/internal/Timeout.java

```diff
@@ -0,0 +1,256 @@
+     *         this class interprets 0 value as 0 elapsed nanoseconds.
+     *     </li>
+     *     <li>
+     *         A positive value means either of the following
+     *         <ol>
+     *             <li>this exact number of nanoseconds has elapsed;</li>
+     *             <li>((n + 1) * 2<sup>63</sup>; (n + 2) * 2<sup>63</sup> - 1]<sup>(*)</sup> nanoseconds has elapsed,
+     *             in which case the timeout has expired.</li>
+     *         </ol>
+     *         Since it is impossible to differentiate the former from the latter, and the former is much more likely to happen in practice,
+     *         this class interprets a positive value as the exact number of elapsed nanoseconds.
+     *     </li>
+     * </ul>
+     * <hr>
+     * <sup>(*)</sup> n is positive and odd.
+     */
+    private long elapsedNanos(final long currentNanos) {
+        assertFalse(isInfinite() || isImmediate());
+        return currentNanos - startNanos;
+    }
+
+    /**
+     * Is package-access for the purpose of testing and must not be used for any other purpose outside of this class.
+     * <p>
+     * Returns 0 or a positive value.
+     * 0 means that the timeout has expired.
+     * <p>
+     * Must not be called on {@linkplain #isInfinite() infinite} timeouts.
+     */
+    long remainingNanos(final long currentNanos) {
+        assertFalse(isInfinite());
+        if (isImmediate()) {
+            return 0;
+        } else {
+            long elapsedNanos = elapsedNanos(currentNanos);
+            return elapsedNanos < 0 ? 0 : Math.max(0, durationNanos - elapsedNanos);
+        }
+    }
+
+    /**
+     * Returns 0 or a positive value.
+     * Use {@link #expired(long)} to check if the returned value signifies that a timeout is expired.
+     *
+     * @throws UnsupportedOperationException If the timeout is {@linkplain #isInfinite() infinite}.
+     * @see #remainingNanosOrInfinite()
+     */
+    public long remainingNanos() throws UnsupportedOperationException {
+        if (isInfinite()) {
+            throw new UnsupportedOperationException();
+        }
+        return remainingNanos(System.nanoTime());
+    }
+
+    /**
+     * Returns a negative value for {@linkplain #isInfinite() infinite} timeouts, otherwise 0 or a positive value.
+     * Use {@link #expired(long)} to check if the returned value signifies that a timeout is expired.
+     *
+     * @see #remainingNanos()
+     */
+    public long remainingNanosOrInfinite() {
+        return isInfinite() ? -1 : remainingNanos();
+    }
+
+    /**
+     * @see #expired(long)
+     */
+    public boolean expired() {
+        return expired(remainingNanosOrInfinite());
+    }
+
+    /**
+     * Returns {@code true} if and only if {@code remainingNanos} is 0.
+     *
+     * @see #remainingNanos()
+     * @see #expired()
+     */
+    public static boolean expired(final long remainingNanos) {
```

Comment:
Does this need to be public? Seems out a strange method for external users.

Reply:
While there is an instance method with the name `expired`, which allows us to check if a timeout has expired, that is not enough. At some point we need to use actual amount of the remaining time because this is what Java SE API and some our internal APIs accept. This is why instance methods `remainingNanos`/`remainingNanosOrInfinite` were introduced. And when we work with the remaining time, we need a way to tell whether or not that value represents an expired timeout or not. While based on the documentation of the `remainingNanos...` methods one could compare the number with 0 and if it is 0, then the timeout is considered expired, I think that writing `if (Timeout.expired(remainingNanos))` is better for readability than writing `if (remainingNanos == 0)`. Currently, this method is used [here](https://github.com/mongodb/mongo-java-driver/pull/685/files?file-filters%5B%5D=.java#diff-7f8dc15248ba3fe10795342e55c0525fee969cc73d918141a5da85dbded487fbR277) and [here](https://github.com/mongodb/mongo-java-driver/pull/685/files?file-filters%5B%5D=.java#diff-7f8dc15248ba3fe10795342e55c0525fee969cc73d918141a5da85dbded487fbR853).